### PR TITLE
BEAM-14026 - Fixes bug related to Unnesting nested rows in an array

### PR DESCRIPTION
**Pull request by abhijeet-lele**
This is address issue of 'unnest'ing arrays containing nested rows.

The issue with the way to reproduce the issue  is described in the https://issues.apache.org/jira/browse/BEAM-14026

### CODE REVIEW

File: sdks/java/extensions/sql/build.gradle [hunk 1]
```diff
@@ -206,6 +206,14 @@ task runBasicExample(type: JavaExec) {
   args = ["--runner=DirectRunner"]
 }
 
+// Run basic SQL example
+task runNestedRowInArrayExample(type: JavaExec) {
+  description = "Run basic SQL example"
+  mainClass = "org.apache.beam.sdk.extensions.sql.example.BeamSqlUnnestExample"
+  classpath = sourceSets.main.runtimeClasspath
+  args = ["--runner=DirectRunner"]
+}
```
Comment (TheNeuralBit) [inline sdks/java/extensions/sql/build.gradle:215]: Do we need this example? It looks like the same behavior is verified continuously in [BeamSqlDslUnnestRowsTest.java](https://github.com/apache/beam/pull/16988/files#diff-8c0c041856075a2f10a667e185965b258fb8dd695a9a7ab9e4fdb417499fd424)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java [hunk 2]
```diff
@@ -108,11 +109,32 @@
...
+                List<Object> rowValues = row.getNestedRowBaseValues();
+                if(null != rowValues) {
+                  values.addAll(rowValues);
+                }
+              } else {
+                values.add(this.getBaseValue(i));
+              }
+              return values.stream();
+            }).flatMap(Function.identity()).collect(Collectors.toList());
+  }
```
Comment (TheNeuralBit) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java:131]: I'd prefer not to add this logic to Row's public API. It's specific to the needs in unnest. Would it be possible to just define it as a helper there?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java [hunk 3]
```diff
@@ -537,7 +559,7 @@ static int deepHashCodeForMap(
     }
 
     static boolean deepEqualsForCollection(
-        Collection<Object> a, Collection<Object> b, Schema.FieldType elementType) {
+            Collection<Object> a, Collection<Object> b, Schema.FieldType elementType) {
```
Comment (TheNeuralBit) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java:562]: I think unnecessary whitespace changes like this are what's making our CI unhappy. Can you run spotless to format the code? `./gradlew spotlessApply`

You might also double check there aren't other checkstyle failures: `./gradlew :sdks:java:extensions:sql:checkstyleMain :sdks:java:extensions:sql:checkstyleTest`

**Review by TheNeuralBit — COMMENTED** [review]
I think the overall approach here is a fine short-term fix, but in the long-term we may want to try to re-use some logic from `Select.flattenedSchema` (see [Select.java](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Select.java))

CC: @reuvenlax @ibzib

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java (hunk 2 above)
Comment (abhijeet-lele) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java]: Moved the logic to the BeamUnnestRel.java

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java (hunk 3 above)
Comment (abhijeet-lele) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java]: done

File: sdks/java/extensions/sql/build.gradle (hunk 1 above)
Comment (abhijeet-lele) [reply sdks/java/extensions/sql/build.gradle]: Removed the example.

**Comment by TheNeuralBit** [comment]
Ran this on some internal Google test suites and it looks good. I'll merge when the Beam CI passes

**Review by TheNeuralBit — APPROVED** [review]

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java (hunk 2 above)
Comment (reuvenlax) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java]: Agree - Row is used by Beam users who don't use SQL. We should not be adding SQL-specific code inside Row.

**Pull request merged**